### ISSUE TITLE
Store and expose reconciliation candidate features

### ISSUE THREAD

**Issue by wetneb** [OpenRefine/OpenRefine#3139]
**Is your feature request related to a problem or area of OpenRefine? Please describe.**

Reconciliation services can expose features in addition to the global score that they return for each candidate, in the upcoming version of the specifications:

https://reconciliation-api.github.io/specs/latest/#dfn-matching-feature

At the moment these are ignored by OpenRefine.

**Describe the solution you'd like**

We should store these features in the reconciliation objects. At the moment, we have a feature array `cell.recon.features` which is computed by OpenRefine statically. These features cannot be tweaked to match the needs of the user or the application domain. For instance, they rely on a hard-coded list of English stop-words.

Moreover they are just defined at the `Recon` level, not for each `ReconCandidate`, making it hard to use them to compare multiple reconciliation candidates for the same cell.

On the long term, I would deprecate and remove `cell.recon.features`, and introduce the features computed by the service as `cell.recon.match.features` (and `cell.recon.candidates[i].features` for each candidate).

This would just give access to the features with GREL, which would not be very convenient on its own.
We could think of ways to display these features in the UI (I have been thinking about a sequence of squares, each filled with a shade of grey indicating the strength of the feature).
We could also design a mechanism to train classifiers based on these features to re-compute a global score which would be more helpful than the generic score returned by the service.

**Describe alternatives you've considered**

The featured returned by the service and the ones computed locally could live alongside each other. But I have personally not found these features very useful, and I rarely see them referred to in tutorials.

**Additional context**

PR for the specs: https://github.com/reconciliation-api/specs/pull/38

**Comment by tfmorris** [comment]
These seem very underspecified. Are the IDs fixed for all W3C compliant reconciliation services or are they just arbitrary strings that a service can make up? If they are random strings, how does a client (specifically OpenRefine) understand what they mean and convey that to the user?

**Comment by wetneb** [comment]
The specs do not specify which features the services should return, each service can make up its own. The idea is that good features tend to be domain-specific so it would not be helpful to constrain features to a fixed range in the specs.

Clients can understand what features mean using the `feature_view` endpoint (see https://github.com/reconciliation-api/specs/pull/38).

**Comment by gitonthescene** [comment]
FWIW, I just discovered that these features are ignored when trying to add this functionality to my reconciliation server framework.  I'd be happy to have them exposed to GREL for starters and leave the graphic representation for later.

As for how they might be used, one case I imagined would help with was [discussed here](https://www.wikidata.org/wiki/User_talk:Gettinwikiwidit#OpenRefine_and_spatial_indexing).  So you could do the reconciliation of say lakes based on coordinates but have a feature which computes the Levenshtein distance based on the name of the lake.  You would then be able to filter based on nearness and on those results filter based on Levenshtein distance.

So effectively they can be used for a two(or more)-step reconciliation process.

Granted, you could try to combine both factors into a single scoring system using weighted scores perhaps, but having them separate gives more visibility into the process.  Another alternative is to do the initial reconciliation, extend with data necessary to make the feature calculation and then do the calculation with a GREL expression.  The advantage of features is that this is less work for the end user and the service provider can provide info likely to be helpful for the domain.

Having them available via GREL would allow you to split out the candidates into different rows along with these feature scores to be able to use facets to view them more carefully.

Since features are domain specific, I'd imagine that they'd be described by the service provider and the user would know about them from the provider.  We could facilitate finding this information by providing a url for the service documentation in the manifest and expose that somewhere in OpenRefine.

As for displaying the features, the spec already refers to a [feature_view](https://reconciliation-api.github.io/specs/latest/#service-manifest) url which I imagine would work similarly to the preview service, perhaps even as a tab in that popup.  I think they're more valuable as part of a calculation than for visual inspection, though.

As for referring to features, attaching them to both `match` and `candidates` makes sense and I'd imagine they'd work similar to the way row's cells are accessed.  I.e. with dot notation when possible and with a key index if necessary.

It's worth pointing out that I imagine these are likely most useful for very specific bespoke reconciliation services when designing a custom reconciliation workflow.

My two cents

**Comment by gitonthescene** [comment]
Linking issue which discusses deprecating static features.  https://github.com/OpenRefine/OpenRefine/issues/3629

**Comment by gitonthescene** [comment]
Also, the `description` field in the reconciliation results is also ignored.  My feeling is that this can be dropped because either a preview or following the view link is likely to provide all the context for a given candidate that you might need.

**Comment by wetneb** [comment]
Thanks for these very thorough thoughts!

Concerning `feature_view` our motivation to introduce it was to let services document their features (associate an URL to each feature so that users could visit that to discover how the feature is computed). Perhaps clients want to embed that in an iframe like we do for entity previews, but intuitively I would assume that this sort of UI is less suited for features (I would expect users to view these documentation pages less often, and potentially they would contain more content than what you could display in a small iframe).

Concerning the `description` field, we introduced it because someone requested that they are shown directly in reconciled cells (without the user having to hover any of the candidates to get a preview). I think it would make a lot of sense to implement this, it would significantly speed up reconciling cells manually. The entity preview should be much richer than simply showing a description (the Wikidata service is a bad example for this), so the two should not be redundant.

However as you have noticed these recent changes to the specs haven't made it yet into OpenRefine (and probably not so many more clients / services) and we haven't formally released them in the specs so if you want to tweak the specs this is the perfect time for it.

And of course I am personally very keen to see these things used in OpenRefine!

**Comment by gitonthescene** [comment]
These thoughts sound very different from my intuition outlined above.  I don’t think that puts me in the best position to expand upon them in the specs.  I’m not sure I even understand how the description field is to be used and can’t picture how it would be more useful than the preview screen.  Is the idea that it’s rendered into the cell and more easily browsed?  I would think it would just feel cluttered.

Somewhat related, it would be nice if the Wikidata reconciliation server had a psuedo property for label available data extension.   Like “label:en” and have that use something like the SPARQL label service.  This is most useful when reconciling by identifier since the Q id is opaque.

In general I think data extension should cover giving more context to a given entry.  Candidates can easily be expanded to extra rows and then reconciled by identifier to allow for data extension.

For explicitness, I’ll add a recipe for splitting out candidates to the Wiki.

**Comment by wetneb** [comment]
> Is the idea that it’s rendered into the cell and more easily browsed? I would think it would just feel cluttered.

Yes that was my understanding (https://github.com/reconciliation-api/specs/issues/47)! But I am also slightly concerned about clutter - perhaps it could just be opt-in? I personally would definitely have found that very useful for many recon projects, but I understand it wouldn't be to everyone's taste…

> Somewhat related, it would be nice if the Wikidata reconciliation server had a psuedo property for label available data extension. Like “label:en” and have that use something like the SPARQL label service. This is most useful when reconciling by identifier since the Q id is opaque.

That already exists: `Len`. Documented at https://wikidata.reconci.link/.

**Comment by gitonthescene** [comment]
Awesome!  Thanks.

**Comment by wetneb** [comment]
The descriptions are also exposed in data extension: `Den`.

If we want people to be able to display descriptions directly in reconciled cells, without any extra fetching operation, it is much easier if those descriptions are directly returned by services as part of the recon candidates.

**Comment by gitonthescene** [comment]
To me this doesn’t feel like a huge win especially considering the potential for clutter.  At the very least it seems like this is a workaround for that [issue you linked](https://github.com/reconciliation-api/specs/issues/47).

But mostly this discussion is a distraction from the missing `features` which would save quite a bit more work.

**Comment by gitonthescene** [comment]
I've added a [description](https://github.com/OpenRefine/OpenRefine/wiki/Recipes#13-reconciliation) of how to split out candidates.  It's lengthy but pretty explicit.  I think this gives you maximal power to treat candidates like any other cell.  There's also a description of how to get back to the original table with augmented info.

**Comment by paulgirard** [comment]
Here are a few usage I have on top of my head of such a system.

## candidates review/acceptance by iterative precision decrease
When you align your data to items you often have to deal with data discrepancy: data properties are not always known.
Moreover the data-set you align with also suffer from discrepancies but not on the same row/properties for more pleasure.
Having access to features scoring in OR would help by allowing to decide to accept candidates which different strategies depending on data availability on both sides (local/distant).
To do so having the individual facets on feature scores would be great to identify the different cases of matching situation.
I stated my example with the binary data presence example but this is also relevant for more complex situation of values comparison.

## recon record
I agree with @gitonthescene having somehting like a way to divide candidates into rows into a record whose first row is original data would be neat.
For each property used by the recon query a new row could propose this feature value in the distant entity + scoring indicators.
such a system would allow to review the scoring "profile" (vectors of feature scores) easily.

## gaining insights also on the data incoming in the recon service
The features score provide more insights on the recon service outputs.
From my experience (see https://github.com/wetneb/openrefine-wikibase/issues/141), it would be useful to help onboard advanced recon services users to provide a log of the recon queries sent by open refine wikibase extension. Indeed the features are actually defined by those queries (I think).
This could also be a specific "preview" panel in the recon modal showing the query template when choosing the recon service options.

## feature scores scales

Feature scores values have different scales depending on scoring algo.
This means that facet on score has to be by feature-score.
This will probably make it difficult for some users to apprehend scores and compare them.
If I think having the "raw" scale is a good thing as it might be necessary for some use case I wonder if proving a normalized score couldn't help for both users and UI design. Such a normalization being not necessarily trivial...

Hope those notes can help.
